### PR TITLE
cleanup / modernize user settings

### PR DESCRIPTION
**Pull request by kdoh**
Makes the user settings page less of a sad place to be. Also removes a bunch of very old one off user settings update code and instead replaces that with the user entity stuff so we keep using / propagating the good stuff. Improves the responsiveness implementation a bit as well. Apologies to the page for the years of neglect.

## New
Screen Shot 2019-05-09 at 5 21 11 PM

## Old
Screen Shot 2019-05-09 at 5 25 09 PM

### Todos
- [x] Remove group selection from the user entity form in this context
- [x] Fix the radio button last child margin issue that's making the tabs slightly off center.

### CODE REVIEW

**Comment by tlrobinson** [comment]
Cool. 

For removing the group field there are a bunch of options:

1. It's pretty easy to create a custom form using the `Form` component directly:

```
<Form
  form={{
    fields: [
      { name: "first_name", ... },
      ...
    ]
  }}
  initialValues={user}
  onSubmit={updatedUser => save the user}
/>
```

We could easily allow you to pass a `form` to `EntityForm` which takes precedence over `entityDef.form`, so you'd automatically get the `onSubmit` etc.

2. We could define multiple named forms in entities, something like:

```
    forms: {
      admin: {
         // what we have now
      },
      user: {
        // same thing without group picker
      }
    }
```

Then have a `formName` prop on `EntityForm` or something (defaulting to the first one if not present?)

3. Third, the fields present in the form can be made dynamic by providing a function that returns an array of fields for the `fields` property (see the `databases` entity for an example). Currently I think we only pass the form values to that function, but we could pass `props` as the second arg or something:

```
form: {
  fields: (values, props) => [
    { name: ... }
  ].concat(!props.isUserForm ?  [
    { name: "group_ids" ... }
  ] : [])
}
```

4. Finally, as you suggested to me on Slack, we could add a field name whitelist or blacklist that to easily exclude fields. One case where this might not work is if you have a top-level `validation` function (as opposed to per-field) on the form that tries to validate a field that you've excluded. I don't think we use the top level one though anywhere (certainly not in user)

**Comment by kdoh** [comment]
@tlrobinson Do you have a favorite of those? Personally I kind of like how explicit 2 is, since then the entityDef itself can let you know that there are multiple types of forms available depending on the situation / entity and the component side stays really simple.

**Comment by camsaul** [comment]
Can we have tests

**Comment by tlrobinson** [comment]
@kdoh Yeah let's go with 2, though 1 is trivial so I'd go ahead and do that too. Here's an implementation of both:

```diff
diff --git a/frontend/src/metabase/entities/containers/EntityForm.jsx b/frontend/src/metabase/entities/containers/EntityForm.jsx
index 56826353e..7fe33efda 100644
--- a/frontend/src/metabase/entities/containers/EntityForm.jsx
+++ b/frontend/src/metabase/entities/containers/EntityForm.jsx
@@ -8,12 +8,25 @@ import ModalContent from "metabase/components/ModalContent";
 
 import entityType from "./EntityType";
 
+function getForm(entityDef, formName = null) {
+  // 1. named form
+  // 2. default `form`
+  // 3. first of the named `forms`
+  return formName
+    ? entityDef.forms[formName]
+    : entityDef.form
+    ? entityDef.form
+    : Object.values(entityDef.forms)[0];
+}
+
 @entityType()
 export default class EntityForm extends React.Component {
   render() {
     const {
       entityDef,
       entityObject,
+      formName,
+      form = getForm(entityDef, formName),
       update,
       create,
       // defaults to `create` or `update` (if an id is present)
@@ -28,7 +41,7 @@ export default class EntityForm extends React.Component {
     const form = (
       <Form
         {...props}
-        form={entityDef.form}
+        form={form}
         initialValues={entityObject}
         onSubmit={onSubmit}
         onSubmitSuccess={action => onSaved && onSaved(action.payload.object)}
```

Not sure if we want to deprecate `form` or leave both, but it's something I'll probably clean up when we extract `entities` into it's own library anyway (technically the form stuff isn't even really related to the rest of `entities`, it's just a convenient place to put it)

**Comment by tlrobinson** [comment]
As far as tests go, I'd take a look at the integration tests at `frontend/test/metabase/admin/people/containers/*.integ.spec.js`

**Comment by mazameli** [comment]
Nice improvement. Some minor nits:

- First name and last name field labels shouldn't have underscores in them
- I'd drop the colons from the labels on the password form fields
- The initials in the circle feel like they could be slightly larger, like maybe the ~16px that the tab labels are. Might also add a smidge of letter-spacing.
- Have you tried using a different accent color for the circle bg? Might be an opportunity to bring in another color from our palette, or to use one of the lighter grays to give it less prominence.

File: frontend/src/metabase/entities/containers/EntityForm.jsx [hunk 1]
```diff
@@ -8,12 +8,28 @@ import ModalContent from "metabase/components/ModalContent";
...
+export function getForm(entityDef: Entity, formName?: any = null) {
+  // 1. named form
+  // 2. default `form`
+  // 3. first of the named `forms`
+  return formName
+    ? // $FlowFixMe
+      entityDef.forms[formName]
+    : entityDef.form
+    ? entityDef.form
+    : Object.values(entityDef.forms)[0];
```
Comment (paulrosenzweig) [inline frontend/src/metabase/entities/containers/EntityForm.jsx]: The third option feels iffy to me. I'm not sure if `Object.values(...)[0]` is guaranteed to return the same value each time. Probably not a real issue if in practice ordering is defined, but it feels weird to rely on ordering.

If we do want to have some form be the default, maybe the entity def should do that manually?

```js
const FORMS = {someForm: {}, otherForm: {}}

const MyEntity = createEntity({
  forms: FORMS,
  form: FORMS.otherForm
})
```

Comment (paulrosenzweig) [reply frontend/src/metabase/entities/containers/EntityForm.jsx]: Looks like my info is outdated and insertion order is preserved: https://stackoverflow.com/a/23202095/607070

Also, I see this behavior is tested below 👍

**Review by paulrosenzweig — APPROVED** [review]

File: frontend/src/metabase/components/form/StandardForm.jsx [hunk 2]
```diff
@@ -34,6 +34,8 @@ const StandardForm = ({
       {form.fields(values).map(formField => {
         const nameComponents = formField.name.split(".");
         const field = getIn(fields, nameComponents);
+
+        console.log("field", field, "name components", nameComponents);
```
Comment (tlrobinson) [inline frontend/src/metabase/components/form/StandardForm.jsx]: remove plz

File: frontend/src/metabase/entities/users.js [hunk 3]
```diff
@@ -21,6 +21,31 @@ export const PASSWORD_RESET_MANUAL =
   "metabase/entities/users/RESET_PASSWORD_MANUAL";
 export const RESEND_INVITE = "metabase/entities/users/RESEND_INVITE";
 
+const BASE_FORM_FIELDS = [
+  // $FlowFixMe - complaining about title not being in object literal?
```
Comment (tlrobinson) [inline frontend/src/metabase/entities/users.js]: typing like

```
import type { FormFieldDef } from "metabase/containers/Form";

const BASE_FORM_FIELDS: FormFieldDef[] = [
```

should do it.

You'll need to

```
export type FormFieldDef
```

in `metabase/containers/Form.jsx`

File: frontend/src/metabase/lib/entities.js [hunk 4]
```diff
@@ -32,6 +32,8 @@ import type { APIMethod } from "metabase/lib/api";
 
 type EntityName = string;
 
+type EntityForm = any;
```
Comment (tlrobinson) [inline frontend/src/metabase/lib/entities.js]: I'd also use the `FormDef` type from `metabase/containers/Form.jsx`

```
import type { FormDef } from "metabase/containers/Form";
```

Like `FormFieldDef` above you'll need to 

```
export type FormDef = {
```

in `metabase/containers/Form.jsx`

**Review by tlrobinson — CHANGES_REQUESTED** [review]
A couple little things but looks good otherwise.

**Review by tlrobinson — APPROVED** [review]

**Pull request merged**